cartridges/openshift-origin-cartridge-jenkins-client/bin/jenkins_build

```diff
@@ -2,37 +2,105 @@
 
 require 'rubygems'
 require 'json'
+require 'net/https'
+require 'uri'
 
 STDOUT.sync = true
 STDERR.sync = true
 Encoding.default_external = "utf-8"
 
+# Build constants from the environment.
 @hostname = ENV['JENKINS_URL'].split("/")[-1]
+@username = ENV['JENKINS_USERNAME']
+@password = ENV['JENKINS_PASSWORD']
 @job_url = "/job/#{ENV['OPENSHIFT_APP_NAME']}-build"
 
+# A list of recoverable HTTP error codes which should indicate
+# a retry is necessary.
+RETRYABLE_ERRORS = ['502', '503', '504']
+
+# Yields until the block stops raising errors. If a retry is
+# attempted, progress and completion is reported; otherwise
+# initial success is silent.
+#
+# Sleeps for a second between retries.
+def try_forever(msg)
+  retried = false
+  loop do
+    begin
+      yield
+      puts "Done." if retried
+      return
+    rescue => e
+      unless retried
+        print "Retrying #{msg}..."
+        retried = true
+      end
+      print '.'
+      sleep 1
+    end
+  end
+end
+
+# Makes an HTTP client with SSL verification disabled.
+def make_http(uri)
+  uri = URI.parse(uri)
+  http = Net::HTTP.new(uri.host, uri.port)
+  http.use_ssl = true
+  http.verify_mode = OpenSSL::SSL::VERIFY_NONE
+  return [http, uri]
+end
+
+# Makes a request to uri and yields the response. Will retry
+# requests whose response code is in RETRYABLE_ERRORS.
+def try_request(type, uri_string, label)
+  http, uri = make_http(uri_string)
+  case type
+  when 'GET'
+    request = Net::HTTP::Get.new(uri.request_uri)
+  when 'POST'
+    request = Net::HTTP::Post.new(uri.request_uri)
+  end
+  request.basic_auth(@username, @password)
+
+  response = nil
+  try_forever(label) do    
+    response = http.request(request)
+    if RETRYABLE_ERRORS.include?(response.code)
+      raise "#{label}: HTTP #{response.code} #{response.message}"
```

Comment:
In this context, I opted to eat it during retries and print periods to indicate progress since the errors are expected.... should I dump the message amongst the progress ellipses every second?